[WIP] Migrations are too aggressive on RenameColumn

Muppets opened the pull request:
Summary of the changes
  - Added foreign key comparison to property diff to ensure foreign key principle table and key match. If not, forces drop and create for the column.
  - Added unit test to cover this case

Fixes #15178

Muppets commented on src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:
```diff
@@ -870,6 +872,16 @@ private static bool EntityTypePathEquals(IEntityType source, IEntityType target,
                    || EntityTypePathEquals(nextSource, nextTarget, diffContext);
         }
 
+        private static bool ForeignKeysMapsToSamePrincipals(IProperty source, IProperty target)
```
Just writing the unit tests to cover these extra scenarios and checking primary key feels wrong. Example scenario:

```
                source =>
                {
                    source.Entity(
                        "Mucor",
                        x =>
                        {
                            x.Property<int>("Id1");
                            x.HasKey("Id1");
                        });
                },
                target =>
                {
                    target.Entity(
                        "Mucor",
                        x =>
                        {
                            x.Property<int>("Id2");
                            x.HasKey("Id2");
                        });
                },
```

To me, this feels like a rename scenario. Is it just the a bad example for primary key or should we not check this?

bricelam replied:
That's a rename scenario--since they're both part of the entity type's primary key, the columns should be paired up. For example, consider this case:
``` csharp
source =>
{
    source.Entity(
        "Mucor",
        x =>
        {
            x.Property<int>("A");
            x.HasKey("A");
        });
},
target =>
{
    target.Entity(
        "Mucor",
        x =>
        {
            x.Property<int>("B");
            x.Property<int>("C");
            x.HasKey("C");
        });
},
```
`A` and `C` should be paired up, but if we don't consider the primary key then `A` and `B` will get paired up instead.

bricelam replied:
Another interesting case:
``` csharp
source =>
{
    source.Entity(
        "Mucor",
        x =>
        {
            x.Property<int>("A");
            x.Property<int>("B");
            x.HasKey("A", "B");
        });
},
target =>
{
    target.Entity(
        "Mucor",
        x =>
        {
            x.Property<int>("A");
            x.HasKey("A");
            x.Property<int>("C");
        });
},
```
`B` and `C` should **not** be paired up because only one is part of the primary key

Muppets replied:
@bricelam Can you have a review of my unit tests? I'm still unsure about where we sit on some of these scenarios. To me, it feels like renaming a primary key column should be a single rename so I don't check the property names of the primary key in the PropertyStructureEquals, just check its the same amount of keys. Feels like a happy medium.

Muppets commented:
@bricelam Assuming you agree with my test scenarios, this is good to go.

bricelam commented:
I haven't had a chance to look at it yet. I'm heads down on EF 6.3. Don't worry, I'll make time before the final EF Core 3.0 release.

bricelam commented on test/EFCore.Relational.Tests/Migrations/Internal/MigrationsModelDifferTest.cs:
```diff
@@ -4225,6 +4225,213 @@ public void Rename_column_with_referencing_foreign_key()
+                    o => Assert.IsType<DropIndexOperation>(o),
+                    o => Assert.IsType<DropColumnOperation>(o),
+                    o => Assert.IsType<AddColumnOperation>(o),
+                    o => Assert.IsType<CreateIndexOperation>(o))
```
This doesn't seem right. I'd expect a column rename.

NeilBostrom replied:
My original thinking on this was if you remove a column pointing at one index then add another pointing at a different one its probably a drop / create wanted. However, thinking on it now, if you rename a column, you would probably rename the index at the same time.

bricelam commented:
@Muppets I think there's still a lot to do on this PR. I want to think through more of the scenarios to decided what all we should check.

If you want to push to get this into 3.0.0, we could scope it back to just the original change (if it's part of a foreign key constraint referencing the same table).

Otherwise, we can take our time on getting the all-up design right and take this in a future release.

Thoughts?

bricelam commented:
(jotted down some notes in https://github.com/bricelam/EFCore/commit/f0b6b0f576da51ec4db8f722347888cc55a5e62e)

NeilBostrom commented:
I'm in no rush on this. Happy for it to wait for another release and get it right.

I'll do some work on those notes and ping something back to you.

bricelam commented:
Ping. I'm ready to look at this again. Did you have any additional changes you'd like to make before I do?

Muppets commented:
I’ve been away from coding for a bit but this has reminded me to get back on it. Give me a week or two to bring this up to date with 3.0 and look at the changes you wanted again.

Muppets commented:
I've included more unit tests to cover the notes from your notes at [bricelam@f0b6b0f](https://github.com/bricelam/EFCore/commit/f0b6b0f576da51ec4db8f722347888cc55a5e62e).

I've added Primary key, foreign key, index and unique constraint checks. I looked at adding check constraint checks too but we don't actually know which columns those are on so can't drill into those.

I simplified the checks by just saying it has to be pointing to the same type and have the same count of properties. This seems to cover all the tests I could come up with.

Muppets commented on src/EFCore.Relational/Migrations/Internal/MigrationsModelDiffer.cs:
```diff
@@ -857,18 +857,33 @@ private bool PropertyStructureEquals(IProperty source, IProperty target)
-            var sourceTable = diffContext.FindSourceTable(source);
-            var targetTable = diffContext.FindTargetTable(target);
-
-            if (sourceTable.EntityTypes.Count == 1
```
@bricelam I was concerned removing this check would break other tests but didn't seem to. This short circuit causes issues when checking that a foreign key points at the same table. This check seems to suggest if both point to a single table then it's the same which feels wrong. The code after it ensure the path is the same etc but only if the count was different?

ajcvickers commented:
This PR still needs work and is lower priority than other items we need to complete for 5.0, so unfortunately we are going to hold off on this for now.

bricelam commented:
My apologies, but given how old this PR is, I'm going to close it for now. I'm not confident taking over maintenance of this code, and I haven't had time to bring it to a place where I am. I'll pick it back up when issue #15178 becomes higher priority or when I have more time to devote to it.
